File: android/src/main/kotlin/com/appodeal/appodeal_flutter/AppodealConsentManager.kt
package com.appodeal.appodeal_flutter

import com.appodeal.ads.Appodeal
import com.appodeal.consent.ConsentForm
import com.appodeal.consent.ConsentInfoUpdateCallback
import com.appodeal.consent.ConsentManager
import com.appodeal.consent.ConsentManagerError
import com.appodeal.consent.ConsentUpdateRequestParameters
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal class AppodealConsentManager(
    private val flutterPlugin: AppodealBaseFlutterPlugin,
    flutterPluginBinding: FlutterPlugin.FlutterPluginBinding
) : MethodChannel.MethodCallHandler {

    val adChannel: MethodChannel = MethodChannel(
        flutterPluginBinding.binaryMessenger,
        "appodeal_flutter/consent_manager"
    ).apply { setMethodCallHandler(this@AppodealConsentManager) }

    private var consentForm: ConsentForm? = null

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "load" -> load(call, result)
            "show" -> show(call, result)
            "loadAndShowIfRequired" -> loadAndShowIfRequired(call, result)
            "revoke" -> revoke(call, result)
            else -> result.notImplemented()
        }
    }

    private fun load(call: MethodCall, result: MethodChannel.Result) {
        val activity = flutterPlugin.activity
        val args = call.arguments as Map<*, *>
        val appKey = args["appKey"] as String
        val tagForUnderAgeOfConsent = args["tagForUnderAgeOfConsent"] as Boolean

        val parameters = ConsentUpdateRequestParameters(
            activity = activity,
            key = appKey,
            tagForUnderAgeOfConsent = tagForUnderAgeOfConsent,
            sdk = "appodeal",
            sdkVersion = Appodeal.getVersion(),
        )

        val infoUpdateCallback = object : ConsentInfoUpdateCallback {
            override fun onFailed(error: ConsentManagerError) {
                val invokeArgs = mapOf("error" to error.localizedMessage)
                adChannel.invokeMethod("onConsentFormLoadFailure", invokeArgs)
            }

            override fun onUpdated() {
                ConsentManager.load(
                    context = activity,
                    successListener = { consentForm ->
                        this@AppodealConsentManager.consentForm = consentForm
                        val status = ConsentManager.status
                        val invokeArgs = mapOf("status" to status.ordinal)
                        adChannel.invokeMethod("onConsentFormLoadSuccess", invokeArgs)
                    },
                    failureListener = { error ->
                        val invokeArgs = mapOf("error" to error.localizedMessage)
                        adChannel.invokeMethod("onConsentFormLoadFailure", invokeArgs)
                    }
                )
            }
        }
        ConsentManager.requestConsentInfoUpdate(parameters, infoUpdateCallback)
        result.success(null)
    }

    private fun show(call: MethodCall, result: MethodChannel.Result) {
        val consentForm = this.consentForm
        if (consentForm == null) {
            val invokeArgs = mapOf("error" to "Consent form is not loaded")
            adChannel.invokeMethod("onConsentFormDismissed", invokeArgs)
        } else {
            val activity = flutterPlugin.activity
            consentForm.show(activity) { error ->
                this.consentForm = null
                val invokeArgs = error?.let { mapOf("error" to error.localizedMessage) }
                adChannel.invokeMethod("onConsentFormDismissed", invokeArgs)
            }
        }
        result.success(null)
    }

    private fun loadAndShowIfRequired(call: MethodCall, result: MethodChannel.Result) {
        val activity = flutterPlugin.activity
        val args = call.arguments as Map<*, *>
        val appKey = args["appKey"] as String
        val tagForUnderAgeOfConsent = args["tagForUnderAgeOfConsent"] as Boolean

        val parameters = ConsentUpdateRequestParameters(
            activity = activity,
            key = appKey,
            tagForUnderAgeOfConsent = tagForUnderAgeOfConsent,
            sdk = "appodeal",
            sdkVersion = Appodeal.getVersion(),
        )

        val infoUpdateCallback = object : ConsentInfoUpdateCallback {
            override fun onFailed(error: ConsentManagerError) {
                val invokeArgs = mapOf("error" to error.localizedMessage)
                adChannel.invokeMethod("onConsentFormLoadFailure", invokeArgs)
            }

            override fun onUpdated() {
                ConsentManager.loadAndShowConsentFormIfRequired(activity) { error ->
                    val invokeArgs = error?.let { mapOf("error" to error.localizedMessage) }
                    adChannel.invokeMethod("onConsentFormDismissed", invokeArgs)
                }
            }
        }
        ConsentManager.requestConsentInfoUpdate(parameters, infoUpdateCallback)
        result.success(null)
    }

    private fun revoke(call: MethodCall, result: MethodChannel.Result) {
        ConsentManager.revoke(flutterPlugin.context)
        result.success(null)
    }
}

File: ios/Classes/AppodealConsentManager.swift
import Flutter
import Foundation
import Appodeal
import StackConsentManager

internal final class AppodealConsentManager {
    
    internal let adChannel: FlutterMethodChannel
    private var dialog: ConsentDialog?
    
    init(registrar: FlutterPluginRegistrar) {
        adChannel = FlutterMethodChannel(name: "appodeal_flutter/consent_manager", binaryMessenger: registrar.messenger())
        adChannel.setMethodCallHandler(handle)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "load": load(call, result)
        case "show": show(call, result)
        case "loadAndShowIfRequired": loadAndShowIfRequired(call, result)
        case "revoke": revoke(call, result)
        case "getPrivacyOptionsRequirementStatus": getPrivacyOptionsRequirementStatus(call, result)
        case "showPrivacyOptionsForm": showPrivacyOptionsForm(call, result)
        default: result(FlutterMethodNotImplemented)
        }
    }
    
    private func load(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let appKey = args["appKey"] as! String
        let tagForUnderAgeOfConsent = args["tagForUnderAgeOfConsent"] as! Bool
        
        let parameters = ConsentUpdateRequestParameters(
            appKey: appKey,
            mediationSdkName: "appodeal",
            mediationSdkVersion: Appodeal.getVersion(),
            COPPA: tagForUnderAgeOfConsent
        )
        
        ConsentManager.shared.requestConsentInfoUpdate(parameters: parameters) { error in
            if let error = error {
                let args: [String: String] = ["error": error.localizedDescription]
                self.adChannel.invokeMethod("onConsentFormLoadFailure", arguments: args)
            } else {
                ConsentManager.shared.load { dialog, error in
                    if let error = error {
                        let args: [String: String] = ["error": error.localizedDescription]
                        self.adChannel.invokeMethod("onConsentFormLoadFailure", arguments: args)
                    } else {
                        self.dialog = dialog
                        let status = ConsentManager.shared.status
                        let args: [String: UInt] = ["status": status.rawValue]
                        self.adChannel.invokeMethod("onConsentFormLoadSuccess", arguments: args)
                    }
                }
            }
        }
        result(nil)
    }
    
    private func show(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let rootViewController = UIApplication.shared.keyWindow?.rootViewController else {
            let args: [String: String] = ["error": "Root view controller is nil"]
            self.adChannel.invokeMethod("onConsentFormDismissed", arguments: args)
            return
        }
        
        if let dialog = self.dialog {
            dialog.present(rootViewController: rootViewController, completion: { error in
                self.dialog = nil
                if let error = error {
                    let args: [String: String] = ["error": error.localizedDescription]
                    self.adChannel.invokeMethod("onConsentFormDismissed", arguments: args)
                } else {
                    self.adChannel.invokeMethod("onConsentFormDismissed", arguments: nil)
                }
            })
        } else {
            let args: [String: String] = ["error": "Consent form is not loaded"]
            self.adChannel.invokeMethod("onConsentFormDismissed", arguments: args)
        }
        result(nil)
    }
    
    
    private func loadAndShowIfRequired(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let appKey = args["appKey"] as! String
        let tagForUnderAgeOfConsent = args["tagForUnderAgeOfConsent"] as! Bool
        
        let parameters = ConsentUpdateRequestParameters(
            appKey: appKey,
            mediationSdkName: "appodeal",
            mediationSdkVersion: Appodeal.getVersion(),
            COPPA: tagForUnderAgeOfConsent
        )
        
        ConsentManager.shared.requestConsentInfoUpdate(parameters: parameters) { error in
            if let error = error {
                let args: [String: String] = ["error": error.localizedDescription]
                self.adChannel.invokeMethod("onConsentFormDismissed", arguments: args)
            } else {
                guard let rootViewController = UIApplication.shared.keyWindow?.rootViewController else {
                    let args: [String: String] = ["error": "Root view controller is nil"]
                    self.adChannel.invokeMethod("onConsentFormDismissed", arguments: args)
                    return
                }
                
                ConsentManager.shared.loadAndPresentIfNeeded(rootViewController: rootViewController) { error in
                    if let error = error {
                        let args: [String: String] = ["error": error.localizedDescription]
                        self.adChannel.invokeMethod("onConsentFormDismissed", arguments: args)
                    } else {
                        self.adChannel.invokeMethod("onConsentFormDismissed", arguments: nil)
                    }
                }
            }
        }
        
        result(nil)
    }
    
    private func revoke(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        ConsentManager.shared.revoke()
        result(nil)
    }

    private func getPrivacyOptionsRequirementStatus(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        // Normalized to match the Dart PrivacyOptionsRequirementStatus enum and the Android side:
        // 0 = unknown, 1 = required, 2 = notRequired.
        let normalized: Int
        switch ConsentManager.shared.privacyOptionsRequirementStatus {
        case .required: normalized = 1
        case .notRequired: normalized = 2
        default: normalized = 0
        }
        result(normalized)
    }

    private func showPrivacyOptionsForm(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let rootViewController = UIApplication.shared.keyWindow?.rootViewController else {
            let args: [String: String] = ["error": "Root view controller is nil"]
            self.adChannel.invokeMethod("onConsentFormDismissed", arguments: args)
            return
        }

        ConsentManager.shared.showPrivacyOptionsForm(rootViewController: rootViewController) { error in
            if let error = error {
                let args: [String: String] = ["error": error.localizedDescription]
                self.adChannel.invokeMethod("onConsentFormDismissed", arguments: args)
            } else {
                self.adChannel.invokeMethod("onConsentFormDismissed", arguments: nil)
            }
        }
        result(nil)
    }
}
